spec/unit/provider/package/puppet_gem_spec.rb

```diff
@@ -20,42 +20,22 @@
     let(:provider_gem_cmd) { '/opt/puppetlabs/puppet/bin/gem' }
   end
 
-  custom_environment = {"HOME"=>ENV["HOME"]}
-  custom_environment['PKG_CONFIG_PATH'] = '/opt/puppetlabs/puppet/lib/pkgconfig'
-
-  let(:execute_options) { {:failonfail => true, :combine => true, :custom_environment => custom_environment} }
+  let(:execute_options) do
+    {
+      failonfail: true,
+      combine: true,
+      custom_environment: {
+        'HOME'=>ENV['HOME'],
+        'PKG_CONFIG_PATH' => '/opt/puppetlabs/puppet/lib/pkgconfig'
+      }
+    }
+  end
 
   before :each do
     resource.provider = provider
-    allow(described_class).to receive(:command).with(:gemcmd).and_return(provider_gem_cmd)
-    allow(Puppet::Util::Platform).to receive(:windows?).and_return(false)
-  end
-
-
-  describe '.windows_gemcmd' do
-    context 'when PUPPET_DIR is not set' do
-      before do
-        # allow(ENV).to receive(:fetch, anything).and_call_original
-        allow(ENV).to receive(:fetch).with('PUPPET_DIR', anything).and_return(nil)
-        allow(Gem).to receive(:default_bindir).and_return('default_gem_bin')
-      end
-
-      it 'uses Gem.default_bindir' do
-        expected_path = File.join('default_gem_bin', 'gem.bat')
-        expect(described_class.windows_gemcmd).to eql(expected_path)
-      end
-    end
-
-    context 'when PUPPET_DIR is set' do
-      before do
-        allow(ENV).to receive(:fetch).with('PUPPET_DIR', anything).and_return('puppet_dir')
-      end
-
-      it 'uses Gem.default_bindir' do
-        expected_path = File.join('puppet_dir', 'bin', 'gem.bat')
-        expect(described_class.windows_gemcmd).to eql(expected_path)
-      end
-    end
+    allow(provider.class).to receive(:which).with(provider_gem_cmd).and_return(provider_gem_cmd)
+    allow(Puppet.run_mode).to receive(:gem_cmd).and_return(provider_gem_cmd)
```

Comment:
As a rule of thumb, I avoid stubbing out the class I'm testing (including code that is inherited or mixed in). Otherwise, the tests can pass for the wrong reason. For example, if I stub out what `Provider#command` returns, then I could have a typo in the `puppet_gem` provider like:

```ruby
command :gemcmd => "/whatever"
```

and the test would still pass.

But the provider/command code is really complicated due to the load time vs runtime time behaviors you mention. So sometimes stubbing the `command` method is a necessary evil.

Reply:
Yes, stubbing and mocking in general are tools I love to avoid when possible. Here it really feels like you can't avoid it. Perhaps a separate test to assert it's not nil without mocking would help?